orca/orca-spec.md

```diff
@@ -0,0 +1,278 @@
+// costs for backends that do not provide a steady stream of telemetry such as
+// long running stream operations and zero QPS services. This is a server
+// streaming service, client needs to terminate current RPC and initiate
+// a new call to change backend reporting frequency.
+service OpenRcaService {
+  rpc StreamCoreMetrics(OrcaLoadReportRequest) returns (stream xds.data.orca.v3.OrcaLoadReport);
+}
+
+message OrcaLoadReportRequest {
+  // Interval for generating Open RCA core metric responses.
+  google.protobuf.Duration report_interval = 1;
+  // Request costs to collect. If this is empty, all known requests costs tracked by
+  // the load reporting agent will be returned. This provides an opportunity for
+  // the client to selectively obtain a subset of tracked costs.
+  repeated string request_cost_names = 2;
+}
+```
+
+The definition for this service can be found at
+https://github.com/cncf/xds/blob/main/xds/service/orca/v3/orca.proto.
+
+Alternatively, an HTTP endpoint can be exposed by the load reporting server. A GET or HEAD request
+to this endpoint will return a response with the `endpoint-load-metrics` header that contains the
+load report in one of the formats listed above.
+
+# ORCA Integration Considerations
+
+## Sidecar
+
+A sidecar, e.g. Envoy or some other in-container proxy process, can inject the inline ORCA header
+at response completion.
+
+![Sidecar Integration Diagram](./orca-spec-sidecar.jpg)
+
+The advantage of a sidecar model is that the application requires no modification as in the example
+service A above. A disadvantage of this approach is that it requires changes to how DPLB operators
+operationally manage their backends, introducing a binary that will require upgrades and version
+management. This might not be a concern in a service mesh architecture where a sidecar can be taken
+as given.
+
+Service B above links with an observability library such as OpenTelemetry that provides support for
+custom metrics. This enables the sidecar to collect additional telemetry data for the ORCA reports.
+
+## Library
+
+The application will link in a library for ORCA that will assist with generating the inline response
+header.
+
+![Library Integration Diagram](./orca-spec-library.jpg)
+
+This will require DPLB operators to modify their applications to call the ORCA library and inject
+the results into the HTTP response.
+
+The library will provide an interface such as:
+
+`std::string GetLoadReportHeaderValue(const std::unordered_map<std::string, double>& named_metrics);`
+
+This method will combine the core metrics such as CPU and Memory (computed by the library) and,
+together with the extra application-specific metrics, compute the inline ORCA response header. The
+application will then be responsible for including this in its response.
+
+# Known DPLB Implementations
+
+## gRPC
+TODO: [Repository]()
+
+[Design](https://github.com/grpc/proposal/blob/master/A51-custom-backend-metrics.md)
```

Comment:
It may also be useful to refer to the following designs as well:

- [gRFC A58: Client-Side Weighted Round Robin](https://github.com/grpc/proposal/blob/master/A58-client-side-weighted-round-robin-lb-policy.md)
- [gRFC A64: xDS LRS Custom Metrics Support](https://github.com/grpc/proposal/blob/master/A64-lrs-custom-metrics.md)
- (pending) [gRFC A85: Changes to xDS LRS Custom Metrics Support](https://github.com/grpc/proposal/pull/454)